### ISSUE TITLE
Dropbox setup with oC never succeeds (oAuth2)

### ISSUE THREAD

**Issue by mmattel** [owncloud/files_external_dropbox#13]
As discussed here: https://central.owncloud.org/t/dropbox-v2-integration-not-working/9795
the Dropbox setup does not succeed.
Following the procedure given from https://owncloud.org/blog/external-storage-dropbox/
everything seems to be setup and entered correct.
You can see the mount in the files app marked with a problem
You can delete the mount
You can try to reset it up without success
No log entries in owncloud.log

**Comment by patrickjahns** [comment]
Can you provide more information? 
What owncloud version are you using?

**Comment by tomneedham** [comment]
@mmattel please check your browser network tab to see the responses from your owncloud server - these give some good feedback to what is happening.

**Comment by mmattel** [comment]
oC 10.0.3, Ubuntu 16.04, php7, nginx, mysql
Client W7x64, Opera

Browser network response (I have replaced sensitive data by xyz):

Step 2 failed. Exception: Client error response [url] https://api.dropboxapi.com/oauth2/token?code=xyz&grant_type=authorization_code&client_id=xyz&client_secret=xyz&redirect_uri=https%3A%2F%2Fxyz%2Fsettings%2Fadmin%3Fsectionid%3Dstorage [status code] 400 [reason phrase] Bad Request"

**Comment by mmattel** [comment]
I did some more digging and it seems that I found the cause.
When looking into the oauth2 client response, I particualry looked if the data entered regarding client_id and client_secret were exact as on the dropbox page. Astonishing the client_secret was identical to the client_id. Than I copied the client_secret to a text editor to see if the copied text was correct and from there I copied it into oC. That worked.
It would be good if the data entered there is not "stared" during the entering phase but when connected it should. I would have identified much earlier any inconsistencies.

There must be something strange because I am no the only one who had exatly this problem and I can´t believe we all are unable to do a proper copy/paste...

**Comment by neemzy** [comment]
As I stated in the OC Central thread, if you check https://github.com/Hemant-Mann/files_external_dropbox/blob/master/lib/Storage/Dropbox.php4 you'll see the constructor expects a configured parameter to be set to 'true', which isn't the case when I check my network tab as suggested by @tomneedham.

Please let me know if I can help any further!

**Comment by patrickjahns** [comment]
@neemzy 
I have tried to replicate the issue and achieved that invalid state but only by chance. I could not reliably reproduce that the XHR Request towards `apps/files_external/globalstorages` would not contain the configured parameter to be defined.

Can you provide steps to reproduce this issue? (What browser are you using?)

@mmattel 
So I tried a couple of times to copy + paste the tokens, and it seems that quite often the "App Secret" would contain white spaces - even though it seems I have just marked the App Secret 😕

**Comment by neemzy** [comment]
I tried with both Firefox and Chrome stable up-to-date. As I just mentioned on OC Central, I'm using CloudFlare for HTTPS. Could that be an issue?

**Comment by mmattel** [comment]
@patrickjahns 
To overcome that issue, it would be good as written above to have the data visible entered into the secret field and star it when connected. You can see the secret at the dropbox page anyway... This would help identify c/p errors.
Q: is it allowed to have white spaces at the secret at the begin or end ?

**Comment by patrickjahns** [comment]
@mmattel 
Might be an idea - the part for entering the credentials is provided by `files_external` app - need to have a look if we can override it specifically for this backend. 

I don't think whitespaces are valid characters - we could discuss if adding a trim enhances the user experiences

**Comment by PVince81** [comment]
Note that there is a likely potential race condition when dealing with OC, not sure if related: https://github.com/owncloud/core/issues/28479

**Comment by patrickjahns** [comment]
@neemzy 
> As I stated in the OC Central thread, if you check https://github.com/Hemant-Mann/files_external_dropbox/blob/master/lib/Storage/Dropbox.php4 you'll see the constructor expects a configured parameter to be set to 'true', which isn't the case when I check my network tab as suggested by @tomneedham.

I've investigated that issue further - and I've seen the following behavior:

Entering some values in the fields `Client ID` and `Client Secret` and waiting for a couple of seconds will trigger a request to `index.php/apps/files_external/globalstorages` that will fail with a 422 and as message
`{"message":"Unsatisfied authentication mechanism parameters"}`. 
Still when pressing the `Grant access` button, I can proceed with the setup (I will be directed towards dropbox website to allow the application to access)

@neemzy 
Can you please tell me, what happens when you press the `Grant access` button? Do you see any javascript errors?

**Comment by neemzy** [comment]
Yes, this is pretty much exactly the error I'm getting clientside (in my case, the endpoint URL ends with `userstorages` but I reckon that's because I'm setting it up for my own user account only and not globally as an admin - and my error message is in French because, well, I'm French, but it's the same). My investigations led me to the aforementioned constructor.

Please note that it only happens when I first enter my DB credentials, or when I toggle the "advanced parameters" menu (with the small gears icon). Clicking the "grant access" button does nothing on my end, including AJAX-wise.

Here's my request payload, if it's of any use:

```
{"mountPoint":"dropbox","backend":"files_external_dropbox","authMechanism":"oauth2::oauth2","backendOptions":{"configured":"","client_id":"p4tndo0z64kk8kj","client_secret":"5tpcft67wfrytpb","token":""},"testOnly":true,"mountOptions":{"encrypt":true,"previews":true,"enable_sharing":false,"filesystem_check_changes":1,"encoding_compatibility":false}}
```

**Comment by henry42** [comment]
I have the same problem, I hack some codes so that the setup can continue. I think if you have the same problem, you can try it temporarily waiting for the update.
1. add `$.getScript("/apps/files_external_dropbox/js/dropbox.js");` at the first line of `apps/files_external/js/dropbox.js` .
2. modify the `apps/files_external_dropbox/vendor/tightenco/collect/src/Illuminate/Support/Collection.php` line 153

    $tmp_values = isset($key) ? $this->pluck($key) : $this;
    $values = $tmp_values->sort()->values();

**Comment by PVince81** [comment]
sorting values??

**Comment by henry42** [comment]
maybe you can ignore this step 2, my php version is 5, if I do not modify `Collection.php`, it prints "PHP Parse error:  syntax error, unexpected '->'"

**Comment by patrickjahns** [comment]
@neemzy 
> Please note that it only happens when I first enter my DB credentials, or when I toggle the "advanced parameters" menu (with the small gears icon). Clicking the "grant access" button does nothing on my end, including AJAX-wise.

The 422 error you are seeing is expected behavior in the sense, it is tested if the parameters entered are correct to report the storage has been setup correctly (translates to red square or green circle). This is the wrong part to focus on.

We need to sort out, why the `grant access` button is not working for you - since this is the crucial step here.
What owncloud version are you using? Do you have any other storages configured?

**Comment by PVince81** [comment]
From reading https://github.com/owncloud/core/issues/28479 it seems that the grant call is done twice in parallel. So on some setups it might depend which one of the calls win, the right one or the wrong one.

**Comment by neemzy** [comment]
> The 422 error you are seeing is expected behavior in the sense, it is tested if the parameters entered are correct to report the storage has been setup correctly (translates to red square or green circle). This is the wrong part to focus on.

I'm not so sure, because whatever triggers the AJAX call, I'd expect it to succeed if my credentials are correct - maybe you mean this specific call doesn't bother including a value for `configured` because it's pretty much expected to fail, but I'd still think it's weird.

However, the "Grant access" button does nothing so far indeed, so it's worth giving a try. My OC version is 10.0.3.3 and no, I have configured no other external storage. FWIW, I have the same issue on both Firefox and Chrome up-to-date (versions 56.0 and 61.0.3163.100 respectively), as I may have stated before. The Dropbox extension version is 1.0.0 (checked on the Market page).

Thank you very much for your help, and please let me know if you need any extra info.

**Comment by henry42** [comment]
@neemzy

have a look at your scripts loaded, in my case `files_external_dropbox/js/dropbox.js` is not loaded so that the button is not triggered

**Comment by patrickjahns** [comment]
@neemzy 
> I'm not so sure, because whatever triggers the AJAX call, I'd expect it to succeed if my credentials are correct - maybe you mean this specific call doesn't bother including a value for configured because it's pretty much expected to fail, but I'd still think it's weird.

The credentials are only part of the setup required for oauth flow to be seen as complete. Once you enter ClientID+Secret you are still required to authorize the application requesting access to your Dropbox. 

When the request is sent just with clientID+Secret it is still missing the token and thus it is expected to fail. So let's now focus on the `grant access` button and figure out why it is not working 


@henry42 
what version of owncloud are you using?

**Comment by neemzy** [comment]
@patrickjahns 

> The credentials are only part of the setup required for oauth flow to be seen as complete. Once you enter ClientID+Secret you are still required to authorize the application requesting access to your Dropbox.

You're absolutely right, got it.
Yes, `dropbox.js` is loaded, [here are its contents from the console](https://dpaste.neemzy.org/paste/59dc9a2c6fe597e049898f2f#javascript) in case that helps.

**Comment by henry42** [comment]
Hi @patrickjahns, mine is 10.0.3 (upgraded from 9.x)

**Comment by henry42** [comment]
@neemzy, it seems that we meet the same problem, it should be "app/files_external_dropbox/js/dropbox.js" not "apps/files_external/js/dropbox.js"

**Comment by neemzy** [comment]
@henry42 Yes indeed, you're right!

**Comment by PVince81** [comment]
Before digging any further, could you guys please try with this core PR: https://github.com/owncloud/core/pull/29145

**Comment by neemzy** [comment]
@PVince81 I just did, it doesn't help with the "Grant access" button doing nothing unfortunately.

**Comment by PVince81** [comment]
@neemzy thanks

**Comment by neemzy** [comment]
@patrickjahns @PVince81 

I think @henry42 has described the issue (wrong JS file loaded) very well; did you have the chance to investigate any further?

**Comment by PVince81** [comment]
According to the code, it should load the correct JS file: https://github.com/owncloud/files_external_dropbox/blob/master/lib/Backend/Dropbox.php#L47

@patrickjahns @davitol can you check in the network console which JS file is loaded for you ?

Can you guys @neemzy confirm that the Dropbox backend you have selected is indeed the one from this app, not the old one from core ? (not sure if the "v2" suffix is present in the released version yet)

**Comment by patrickjahns** [comment]
@neemzy 
There should be two dropbox.js files served - one from `apps/files_external/js/dropbox.js` (the one that is legacy v1 and part of files_external) and the other one via `custom/files_external_dropbox/js/dropbox.js` (which is from this bundle).

It is registered here: https://github.com/owncloud/files_external_dropbox/blob/master/lib/Backend/Dropbox.php#L47

**Comment by neemzy** [comment]
@PVince81 Yes, I'm positive it is the right one, I disabled the legacy one from the admin panel to avoid making this mistake.

@patrickjahns I checked again and I only get the one from `files_external`.

**Comment by PVince81** [comment]
@neemzy so you are mounting Dropbox as a user, not admin, is that correct ? You unticked the legacy Dropbox from the list of allowed backends for users ?

@patrickjahns @davitol did you guys test with Dropbox mounted as user, not admin ?

**Comment by davitol** [comment]
> @patrickjahns @davitol did you guys test with Dropbox mounted as user, not admin ?

yes i did, there was an issue and it was solved before the release

**Comment by davitol** [comment]
JFYI it is needed to add the personal url in the DropBox app developers console:

http://path/to/owncloud/index.php/settings/personal?sectionid=storage 

not only http://path/to/owncloud/index.php/settings/admin?sectionid=storage

**Comment by neemzy** [comment]
@PVince81 Yes, that is correct, but FYI I tried as an admin too (changing the callback URL in my Dropbox app accordingly) and got the same results. And yes, legacy Dropbox has been unticked as soon as I installed the extension for the new version, to avoid any mistake on my end.

**Comment by henry42** [comment]
@neemzy  @PVince81  @patrickjahns
I finally found the reason, and create a pull request

**Comment by neemzy** [comment]
@PVince81 Yes, that is correct, but FYI I tried as an admin too (changing the callback URL in my Dropbox app accordingly) and got the same results. And yes, legacy Dropbox has been unticked as soon as I installed the extension for the new version, to avoid any mistake on my end.

**Comment by PVince81** [comment]
@henry42 great find! This will likely affect other external storages as well.

**Comment by PVince81** [comment]
PR here: https://github.com/owncloud/core/pull/29217

Those with the issue, please try it out!

**Comment by neemzy** [comment]
Applying the changes made the "Grant access button" work and the external storage configuration apparently succeed. The configuration row was duplicated in the process somehow, but I just had to delete the duplicate.

Just a small issue: my "files" page is now empty. I mean, there's the menu on the left side, but the file list is empty. Even after removing the external storage. That's kinda inconvenient.

**Comment by neemzy** [comment]
Setting up the external storage now works, but the file list is broken (empty) until I remove it.
In the console, I can see a 500 response to a PROPFIND request to `https://owncloud.url/remote.php/webdav/`, and a JS error stating that `xml is null` (`var xml = response.xhr.responseXML;` in `client.js`).

**Comment by PVince81** [comment]
@neemzy please check the owncloud.log and error log, there should hopefully be an entry that matches this 500 error.

**Comment by neemzy** [comment]
@PVince81 Yup.

`[Thu Oct 12 10:09:34.216917 2017] [:error] [pid 20347] [client 108.162.229.238:22565] PHP Parse error:  syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /path/to/owncloud/apps/files_external_dropbox/vendor/tightenco/collect/src/Illuminate/Support/Collection.php on line 154`

**Comment by PVince81** [comment]
Argh, it seems this is a third party library and composer doesn't enforce PHP 5.6. I raised a separate ticket for that: https://github.com/owncloud/files_external_dropbox/issues/14

**Comment by neemzy** [comment]
I reckon we can close this issue if my remaining problem is unrelated and @henry42's PR works for other people :)

**Comment by Hemant-Mann** [comment]
@neemzy could you share your composer.lock file that composer generated?

**Comment by patrickjahns** [comment]
@Hemant-Mann 
The composer.lock file is inside the repository - but it was created on a php7.1 environment (my environment) since it was missing. We need to re-create it on php5.6 and go through QA

**Comment by Hemant-Mann** [comment]
Yes I think you are right
the version specified in the file - https://github.com/owncloud/files_external_dropbox/blob/master/composer.lock#L503 is only supported by php >= 7
see https://packagist.org/packages/tightenco/collect#v5.5.13

**Comment by patrickjahns** [comment]
1.0.1 is in marketplace - issue should be resolved
